### PR TITLE
mitosheet: add ending condition to Excel Range Import

### PR DESCRIPTION
**Pull request by naterush**
# Description

Adds Excel Range Import changes per spec. Note how it additionally makes the generated code dynamic for the find by specific value, as this is also a requirement.

# Testing

Test the use case we're targeting!

# Documentation

No.

### CODE REVIEW

File: mitosheet/mitosheet/public/v2/__init__.py [hunk 1]
```diff
@@ -5,4 +5,5 @@
     FUNCTIONS
 )
 from mitosheet.public.v1.sheet_functions import *
-from mitosheet.public.v1 import register_analysis
\ No newline at end of file
+from mitosheet.public.v1 import register_analysis
+from mitosheet.public.v2.excel_utils import get_table_range_from_upper_left_corner_value, get_read_excel_params_from_range
```
Comment (naterush) [inline mitosheet/mitosheet/public/v2/__init__.py:9]: Note that we add to the public interface

File: mitosheet/mitosheet/public/v2/excel_utils.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def get_table_range_from_upper_left_corner_value(file_path: str, sheet_name: str, upper_left_value: Union[str, int, float, bool], bottom_left_value: Optional[Union[str, int, float, bool]]=None) -> Optional[str]:
+    """
+    Given a string, this function will look through the excel tab sheet_name at the given
+    file_path, and find the first instance of the value (look through column A first, then B, etc).
+
+    If this value does not exist in the tab, will return None. 
+
+    If the value does exist, then this function will walk down the column until it hits an empty row. Then,
```
Comment (naterush) [inline mitosheet/mitosheet/public/v2/excel_utils.py:17]: Update comment

File: mitosheet/mitosheet/tests/step_performers/import_steps/test_excel_range_import.py [hunk 3]
```diff
@@ -33,97 +34,124 @@
...
+        ['A1:B4'],
+        [TEST_DF_3],
+        [{'type': 'upper left corner value', 'end_condition': {'type': 'bottom left corner value', 'value': 200}, 'df_name': 'dataframe_1', 'value': TEST_DF_3.columns[0]}],
+        [TEST_DF_3.iloc[0:2]],
+    ),
+    (
+        ['A1:B2', 'A4:B5'],
+        [pd.DataFrame({'A': [1]}), pd.DataFrame({'B': [100]})],
+        [{'type': 'upper left corner value', 'end_condition': {'type': 'bottom left corner value', 'value': 100}, 'df_name': 'dataframe_1', 'value': 'A'}],
+        [pd.DataFrame({'A': [1, None, 'B', 100]})],
```
Comment (naterush) [inline mitosheet/mitosheet/tests/step_performers/import_steps/test_excel_range_import.py:135]: Try with a string

**Review by naterush — COMMENTED** [review]
Minor cleanup, testing.

**Comment by aarondr77** [comment]
Table end condition text is different from the rest

Screenshot 2023-03-13 at 10 04 02 PM

**Comment by aarondr77** [comment]
When this screen is full, I cannot scroll in the taskpane, but notice that the text at the top is cutoff. 

Screenshot 2023-03-13 at 10 07 14 PM

**Comment by aarondr77** [comment]
1. Open the range import taskpane
2. Configure card A
3. Configure card B
4. Import the data
5. Reconfigure card A
6. Import the data
7. Notice that the data does not change. However, in step 5, if you instead reconfigured card B, it would reflect the update. 



https://user-images.githubusercontent.com/18709905/224875494-576d91bc-73e7-479d-af51-b0b563b70875.mov

**Comment by aarondr77** [comment]
The dynamic code is A++ ! Once we fix test + fix these bugs lgtm with some heavy testing

File: mitosheet/mitosheet/code_chunks/step_performers/import_steps/excel_range_import_code_chunk.py [hunk 4]
```diff
@@ -10,52 +10,87 @@
...
 from mitosheet.excel_utils import (get_column_from_column_index,
                                    get_col_and_row_indexes_from_range)
+from mitosheet.public.v2.excel_utils import get_read_excel_params_from_range
 from mitosheet.state import State
+from mitosheet.transpiler.transpile_utils import column_header_to_transpiled_code
+from mitosheet.types import ExcelRangeImport
+
+
+EXCEL_RANGE_IMPORT_TYPE_RANGE = 'range'
+EXCEL_RANGE_IMPORT_TYPE_UPPER_LEFT = 'upper left corner value'
```
Comment (aarondr77) [inline mitosheet/mitosheet/code_chunks/step_performers/import_steps/excel_range_import_code_chunk.py:20]: EXCEL_RANGE_IMPORT_TYPE_UPPER_LEFT -> EXCEL_RANGE_IMPORT_TYPE_UPPER_LEFT_VALUE for consistency

File: mitosheet/mitosheet/saved_analyses/step_upgraders/excel_range_import.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            "sheet_name": str,
+            "range_imports": {
+                "type": 'range',
+                "df_name": str,
+                "value": str
+            } | {
+                "type": 'upper left corner value',
+                "df_name": str,
+                "value": str,
+                "end_condition": {'type': 'first empty cell'} | {'type': 'bottom left corner value', 'value': Any}
```
Comment (aarondr77) [inline mitosheet/mitosheet/saved_analyses/step_upgraders/excel_range_import.py:45]: How come value is `str` for the upper left corner value, but `Any` for the end condition value?

Comment (aarondr77) [reply mitosheet/mitosheet/saved_analyses/step_upgraders/excel_range_import.py:45]: Oooo its actually not in the type definition: ExcelRangeEndCondition's value is Optional[str]

File: mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx [hunk 6]
```diff
@@ -259,7 +284,94 @@ const ExcelRangeImportTaskpane = (props: ExcelRangeImportTaskpaneProps): JSX.Ele
...
                                 </Col>
                             </Row>
+                            {range_import.type === 'upper left corner value' && 
+                                <>
+                                    <Row justify="space-between">
+                                        <Col>
+                                            <p>
+                                                Table End Condition
+                                            </p>
+                                        </Col>
```
Comment (aarondr77) [inline mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx:294]: To keep the formatting consistent, we could make this a label and tooltip component and add some helpful tooltip

File: mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx [hunk 7]
```diff
@@ -259,7 +284,94 @@ const ExcelRangeImportTaskpane = (props: ExcelRangeImportTaskpaneProps): JSX.Ele
...
+                                        <Col>
+                                            <Select
+                                                width="medium"
+                                                value={range_import.end_condition.type}
+                                                onChange={(newType) => {
+                                                    setParams((prevParams) => {
+                                                        const newRangeImports = window.structuredClone(prevParams.range_imports);
+                                                        const newRangeImport: ExcelRangeUpperLeftCornerValueImport = window.structuredClone(range_import);
+
+                                                        if (newType === 'first empty cell') {
```
Comment (aarondr77) [inline mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx:304]: Maybe we can first cast `newType` to ExcelRangeImportType so we make sure these string checks remain consistent if we change the ExcelRangeImportType options. 


Can we also create a test to make sure that its kept in sync with the backend

**Review by aarondr77 — COMMENTED** [review]
LGTM after a bit more full integration testing once these bugs are fixed.

File: mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx (hunk 6 above)
Comment (naterush) [reply mitosheet/src/components/taskpanes/ExcelRangeImport/ExcelRangeImportTaskpane.tsx:294]: I think the question marks might be just additional noise in this case. The dropdown has a ton of information in it. I made the text consistent though.

File: mitosheet/mitosheet/saved_analyses/step_upgraders/excel_range_import.py (hunk 5 above)
Comment (naterush) [reply mitosheet/mitosheet/saved_analyses/step_upgraders/excel_range_import.py:45]: Yeah. It's the best we can do in MyPy I think.
